src/Classes/SkillsTab.lua

```diff
@@ -860,6 +868,19 @@ function SkillsTabClass:FindSkillGem(nameSpec)
 	return "Unrecognised gem name '"..nameSpec.."'"
 end
 
+function SkillsTabClass:MatchGemLevelToCharacterLevel(gemData, fallbackGemLevel)
+	if self.matchGemLevelToCharacterLevel then
+		local maxGemLevel = m_min(self.defaultGemLevel or gemData.defaultLevel, gemData.defaultLevel + 1)
+		for gemLevel = maxGemLevel, 1, -1 do
+			if gemData.grantedEffect.levels[gemLevel] and gemData.grantedEffect.levels[gemLevel].levelRequirement <= self.build.characterLevel then
```

Comment:
You can shortcut this loop a bit by skipping to the maximum level that does exist if it doesn't.  e.g.

```lua
if not gemData.grantedEffect.levels[maxGemLevel] then
    maxGemLevel = #gemData.grantedEffect.levels
end
```

and do your for loop from there minus the initial if expression (actually you might still need that if this affects skills from items